<h1>Atividade - Agrupamentos e Dispersão</h1>

<p>=============================================================================================================</p>

- Utilizando a base em anexo (glp-2021-02.csv):
<ol>
    <li>Análise completa sobre o describe</li>
    <li>Fazer a amplitude interquartil do Valor de Venda e uma análise sobre o resultado apresentado</li>
    <li>Fazer o coeficiente de variação e uma análise sobre ele</li>
    <li>Fazer o agrupamento das top 10 bandeiras com a média do maior valor de venda</li>
    <li>Fazer o agrupamento por estado e trazer os 5 estados com a média do maior valor de venda e os 5 estados com a média do menor valor de venda</li>
    <li>Trazer o agrupamento por mês com a média de valor de venda por mês</li> 
</ol>






<p>=============================================================================================================</p>

<h3>Importando as bibliotecas</h3>

In [662]:
import pandas as pd
import numpy as np

<h3>Lendo o arquivo .csv com pandas</h3>

In [663]:
df = pd.read_csv('glp-2021-02.csv', delimiter=';')

# Excluindo a coluna Valor de Compra
#df = df.drop(columns=['Valor de Compra'])
#df = df.rename(columns={'Regiao - Sigla':'Regiao_Sigla', 'Estado - Sigla':'Estado_Sigla'})
df.head()

Regiao - Sigla Estado - Sigla            Municipio  \
0             NE             CE            MARACANAU   
1              S             SC       JARAGUA DO SUL   
2             NE             MA  SAO JOSE DE RIBAMAR   
3             SE             SP              BIRIGUI   
4             SE             SP                GARCA   

                                Revenda      CNPJ da Revenda  \
0                       POSTO BOTO LTDA   03.020.995/0001-40   
1                     POSTO CIDADE LTDA   81.352.882/0004-07   
2              POSTO ATLANTICO SUL LTDA   04.034.035/0001-00   
3  AUTO POSTO SCUCUGLIA E OLIVEIRA LTDA   00.226.807/0001-47   
4          AUTO POSTO RANELI GARCA LTDA   72.812.795/0001-43   

                          Nome da Rua Numero Rua Complemento  \
0  AVENIDA PADRE JOSE HOLANDA DO VALE        809         NaN   
1                RUA WALTER MARQUARDT       2555         NaN   
2                     RODOVIA MA  203         20         K M   
3                      RUA CONSOLACAO        140         NaN   
4                  RUA BASILIO SELANI         45         NaN   

               Bairro        Cep Produto Data da Coleta Valor de Venda  \
0         PIRATININGA  61905-000     GLP     01/07/2021          89,99   
1  BARRA DO RIO MOLHA  89259-700     GLP     01/07/2021           86,9   
2             ARACAGY  65110-000     GLP     01/07/2021          89,99   
3  PARQUE R PINHEIROS  16203-031     GLP     01/07/2021             85   
4   VILA JOSE RIBEIRO  17400-000     GLP     01/07/2021            100   

   Valor de Compra Unidade de Medida              Bandeira  
0              NaN        R$ / 13 kg   NACIONAL GÁS BUTANO  
1              NaN        R$ / 13 kg   NACIONAL GÁS BUTANO  
2              NaN        R$ / 13 kg              LIQUIGÁS  
3              NaN        R$ / 13 kg  SUPERGASBRAS ENERGIA  
4              NaN        R$ / 13 kg              LIQUIGÁS

In [664]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89372 entries, 0 to 89371
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Regiao - Sigla     89372 non-null  object 
 1   Estado - Sigla     89372 non-null  object 
 2   Municipio          89372 non-null  object 
 3   Revenda            89372 non-null  object 
 4   CNPJ da Revenda    89372 non-null  object 
 5   Nome da Rua        89372 non-null  object 
 6   Numero Rua         89371 non-null  object 
 7   Complemento        21059 non-null  object 
 8   Bairro             89367 non-null  object 
 9   Cep                89372 non-null  object 
 10  Produto            89372 non-null  object 
 11  Data da Coleta     89372 non-null  object 
 12  Valor de Venda     89372 non-null  object 
 13  Valor de Compra    0 non-null      float64
 14  Unidade de Medida  89372 non-null  object 
 15  Bandeira           89372 non-null  object 
dtypes: float64(1), object(

<h3> Tratando a coluna 'Valor de Venda' </h3>

In [665]:
# Trocando , por . da coluna 'Valor de Venda'
# read_csv(decimal=',')
df['Valor de Venda'] = df['Valor de Venda'].str.replace(',', '.')
df['Valor de Venda'].dtypes

dtype('O')

In [666]:
# Convertendo o tipo da coluna 'Valor de Venda' para Float
df['Valor de Venda'] = df['Valor de Venda'].astype(float)
df['Valor de Venda'].dtypes

dtype('float64')

In [667]:
df.head(5)

Regiao - Sigla Estado - Sigla            Municipio  \
0             NE             CE            MARACANAU   
1              S             SC       JARAGUA DO SUL   
2             NE             MA  SAO JOSE DE RIBAMAR   
3             SE             SP              BIRIGUI   
4             SE             SP                GARCA   

                                Revenda      CNPJ da Revenda  \
0                       POSTO BOTO LTDA   03.020.995/0001-40   
1                     POSTO CIDADE LTDA   81.352.882/0004-07   
2              POSTO ATLANTICO SUL LTDA   04.034.035/0001-00   
3  AUTO POSTO SCUCUGLIA E OLIVEIRA LTDA   00.226.807/0001-47   
4          AUTO POSTO RANELI GARCA LTDA   72.812.795/0001-43   

                          Nome da Rua Numero Rua Complemento  \
0  AVENIDA PADRE JOSE HOLANDA DO VALE        809         NaN   
1                RUA WALTER MARQUARDT       2555         NaN   
2                     RODOVIA MA  203         20         K M   
3                      RUA CONSOLACAO        140         NaN   
4                  RUA BASILIO SELANI         45         NaN   

               Bairro        Cep Produto Data da Coleta  Valor de Venda  \
0         PIRATININGA  61905-000     GLP     01/07/2021           89.99   
1  BARRA DO RIO MOLHA  89259-700     GLP     01/07/2021           86.90   
2             ARACAGY  65110-000     GLP     01/07/2021           89.99   
3  PARQUE R PINHEIROS  16203-031     GLP     01/07/2021           85.00   
4   VILA JOSE RIBEIRO  17400-000     GLP     01/07/2021          100.00   

   Valor de Compra Unidade de Medida              Bandeira  
0              NaN        R$ / 13 kg   NACIONAL GÁS BUTANO  
1              NaN        R$ / 13 kg   NACIONAL GÁS BUTANO  
2              NaN        R$ / 13 kg              LIQUIGÁS  
3              NaN        R$ / 13 kg  SUPERGASBRAS ENERGIA  
4              NaN        R$ / 13 kg              LIQUIGÁS

<h3>1. Análise completa sobre o Describe</h3>

In [668]:
df['Valor de Venda'].describe().round(2)

count    89372.00
mean        99.28
std          9.59
min         64.99
25%         93.00
50%         99.00
75%        105.00
max        140.00
Name: Valor de Venda, dtype: float64

In [669]:
df['Valor de Venda'].mode()

0    100.0
dtype: float64

In [670]:
round(df[df['Valor de Venda']==100].count()[-4] / df.count()[0] * 100, 0)

11.0

In [671]:
# df[df['Valor de Venda'] == df['Valor de Venda'].max()][
#     ['Regiao - Sigla', 'Estado - Sigla', 'Municipio', 'Revenda', 'Bairro', 'Valor de Venda', 'Bandeira']].head()

<h3>O gás com o valor mais alto de cada estado</h3>

In [672]:
df.loc[df.groupby('Estado - Sigla')['Valor de Venda'].idxmax()][['Estado - Sigla', 'Bandeira', 'Valor de Venda']]

Estado - Sigla              Bandeira  Valor de Venda
58675             AC                 FOGAS           130.0
36321             AL              LIQUIGÁS           105.0
63683             AM                 FOGAS           125.0
58147             AP              LIQUIGÁS           127.0
48000             BA               BAHIANA           115.0
48837             CE              LIQUIGÁS           115.0
78036             DF                BRANCA           115.0
29882             ES                BRANCA           110.0
39214             GO  SUPERGASBRAS ENERGIA           130.0
72599             MA                BRANCA           120.0
53040             MG  SUPERGASBRAS ENERGIA           125.0
47771             MS               COPAGAZ           124.0
59244             MT  SUPERGASBRAS ENERGIA           140.0
14308             PA   NACIONAL GÁS BUTANO           130.0
50985             PB   NACIONAL GÁS BUTANO           120.0
80222             PE              MINASGAS           120.0
48627             PI              LIQUIGÁS           120.0
57771             PR  SUPERGASBRAS ENERGIA           130.0
59401             RJ   NACIONAL GÁS BUTANO           125.0
49070             RN   NACIONAL GÁS BUTANO           120.0
46158             RO             AMAZONGÁS           135.0
42912             RR             AMAZONGÁS           130.0
64695             RS              ULTRAGAZ           139.0
51324             SC  SUPERGASBRAS ENERGIA           135.0
48526             SE               BAHIANA           100.0
51311             SP              LIQUIGÁS           130.0
36552             TO              LIQUIGÁS           120.0

<p><b>Análise:</b></p>
<ol>
    <li><b>25%</b> da nossa base vende o gás a <b>R&#x24; 93,00</b> ou menos.</li>
    <li><b>50%</b> da nossa base vende o gás a <b>R&#x24; 99,00</b>  ou menos.</li>
    <li><b>75%</b> da nossa base vende o gás a <b>R&#x24; 105,00</b>  ou menos.</li>
    <li>Em média, os revendedores venderam o gás a <b>R&#x24; 99,00</b>  no segundo semestre de 2021.</li>
    <li><b>11%</b> dos revendedores, venderam o gás a <b>R&#x24; 100,00.</b> </li>   
<ol>

<h3>2. Fazer a amplitude interquartil do Valor de Venda e uma análise sobre o resultado apresentado</h3>

In [673]:
# 75%
Q3 = df['Valor de Venda'].describe()[6]
Q3

105.0

In [674]:
# 25%
Q1 = df['Valor de Venda'].describe()[4]
Q1

93.0

In [675]:
# Intervalo = Q3 - Q1
intervalo = Q3 - Q1

In [676]:
# Intervalo Interquartil do Valor de Venda:
print(f'Interquartil do Valor de Venda é: {intervalo}')

Interquartil do Valor de Venda é: 12.0


<p><b>Análise:</b></p>
<ol>
    <li>A amplitude interquartil do valor de venda é de <b>R&#x24; 12,00.</b><br></li>
    <li><b>50%</b> dos dados estão entre os quartis inferior e superior.</li>
    <li>Valor de venda varia no máximo em <b>R&#x24; 105,00.</b></li>
    <li> Os outliers do nosso valor de venda é <b>R&#x24; 64.99</b> (Menor Valor de Venda) e <b>R&#x24;140,00</b> (Maior Valor de Venda).</li>
</ol>

<h3>3. Fazer o coeficiente de variação e uma análise sobre ele</h3>

In [677]:
# Média do Valor de Venda
media_vv = round(df['Valor de Venda'].mean(), 2)
media_vv

99.28

In [678]:
# Desvio Padrão do Valor de Venda
desvio_padrao_vv = round(df['Valor de Venda'].describe()['std'], 2)
desvio_padrao_vv

9.59

In [679]:
# Coeficiente de Variaação
# desvio padrão / média * 100
coeficiente_variacao_vv = round(desvio_padrao_vv / media_vv * 100, 0)

print(f'O Coeficiente do Valor de Venda é {coeficiente_variacao_vv}%')

O Coeficiente do Valor de Venda é 10.0%


<p><b>Análise:</b></p>

<ol>
    <li>O <b>coeficiente de variação</b> é de <b>10%</b> em torno da média, ou seja,<br>
        é uma variação aceitável, pois temos uma grande quantidade de dados agrupados.<br>
        Através dos nossos percentis, podemos perceber o agrupamento dos dados.<br>
        <b>25% -> 93.00</b><br>
        <b>50% -> 99.00</b><br> 
        <b>75% -> 105.00</b>
    </li> 
<ol>

In [680]:
# def agrupamento(gb_coluna, coluna, funcao, qtd_exibir):
#     formula = df.groupby(gb_coluna)[[coluna]].mean().nsmallest(qtd_exibir, coluna).round(2)
#     return formula
# agrupamento('Bandeira', 'Valor de Venda', 5)

<h3>4. Fazer o agrupamento das top 10 bandeiras com a média do maior valor de venda</h3>

In [681]:
df.groupby('Bandeira')[['Valor de Venda']].mean().nlargest(10, 'Valor de Venda').round(2)

Valor de Venda
Bandeira                            
FOGAS                         109.20
AMAZONGÁS                     107.18
SUPERGASBRAS                  107.17
COPAGAZ                       101.46
ULTRAGAZ                      100.10
LIQUIGÁS                      100.09
NACIONAL GÁS BUTANO            99.00
SUPERGASBRAS ENERGIA           98.48
BRANCA                         98.19
SERVGÁS                        96.69

<h3>5.1 Fazer o agrupamento por estado e trazer os 5 estados com a média do maior valor de venda</h3>

In [682]:
df.groupby('Estado - Sigla')[['Valor de Venda']].mean().nlargest(5, 'Valor de Venda').round(2)

Valor de Venda
Estado - Sigla                
MT                      118.69
AC                      113.44
RO                      113.26
AP                      111.88
RR                      110.66

<h3>5.2 Fazer o agrupamento por estado e trazer os 5 estados com a média do menor valor de venda</h3>

In [683]:
df.groupby('Estado - Sigla')[['Valor de Venda']].mean().nsmallest(5, 'Valor de Venda').round(2)

Valor de Venda
Estado - Sigla                
RJ                       89.30
PE                       91.23
BA                       92.19
SE                       92.26
DF                       94.30

<h3>6. Trazer o agrupamento por mês com a média de valor de venda por mês</h3>

In [692]:
# Convertendo o tipo da coluna 'Data da Coleta' para datetime
df['Data da Coleta'] = pd.to_datetime(df['Data da Coleta'])
df['Data da Coleta'].dtypes

dtype('<M8[ns]')

In [693]:
# Pegando apenas o mês da data
data_coleta = df['Data da Coleta'].dt.month

In [695]:
#df.groupby(df['Data da Coleta'].dt.month)[['Valor de Venda']].mean().round(2)
df.groupby(data_coleta)[['Valor de Venda']].mean().round(2)

Valor de Venda
Data da Coleta                
1                        99.10
2                        98.69
3                       100.03
4                        97.94
5                        95.59
6                        97.96
7                        94.58
8                        96.27
9                        99.86
10                      102.27
11                      102.38
12                      102.11